### PR TITLE
Create Application: Improvements October 2020

### PR DESCRIPTION
**Pull request by AleziaKurdis**
ISSUES ADDRESSED:
**Issue #711  Add error dialogue if an unsupported filetype is added as a model URL in the create app.**
Supported type of files are now listed next to the label "Model URL"

**Issue #798 Angular Velocity can't go below 0.4 something (round to 0.5)** 
The precision in the UI has been increased. This doesn't solve the issue but make this more comfortable.
(We rely now on issue #76 to address this problem globally.)

ENTITY LIST UI ENHANCEMENTS:
This PR adds an "Actions" and a "Selection" menu to the Create App. Entity List _(Desktop and HMD compliant)_
Such a menu has now room for more actions than the previous buttons where we can't have tooltip easily in HMD.
It will be easy to add new actions and Selection algorithms in the future. Also no more cryptic pictogram.

**Supported Actions:**
- Undo
- Redo
- Cut
- Copy
- Paste
- Duplicate
- Delete
- Parent Entity to Last Selected
- Unparent Entity
- Move Selected Entities to Avatar
image

**Supported selection operations:**
- Select All
- Select None
- Inverse Selection
- Select All Entities In Box
- Select All Entities Touching Box
- Teleport to Selected Entities
image

**Note:** Previous buttons have been removed. They are now in this menu.
At one exception: the **HMD specific "Multiple Selection"** that has been kept as a button.

PARENT HIERARCHY DISPLAY:
A new optional "**Hierarchy**" column has been added to the Entity List. 
This column shows 4 possible values:
- Parent
- Child
- Parent & Child
- None

image

An entity selected in-world has now an **orange box** display when the entity is a **parent** and a **cyan** one when it is a **child only**. Otherwise, it displays gray as before (including when it is a multiple selection)
The objective here is simply to help, by giving a clue. This could be revisited later if we need a more accurate display.
image
image

### CODE REVIEW

**Comment by AleziaKurdis** [comment]
Do not merge now.
This PR has been committed to be reviewed and tested. (Please do and give me your feedback)
Additional bug fix and enhancements will be added in the next days.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-109804b_affad674.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-109804b.tar_affad674.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-109804b.tar_affad674.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-109804b_affad674.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-109804b_affad674.exe

File: scripts/system/html/css/edit-style.css
Comment (ctrlaltdavid) [inline scripts/system/html/css/edit-style.css:422]: Unexpected comment.

File: scripts/system/html/css/edit-style.css
Comment (ctrlaltdavid) [inline scripts/system/html/css/edit-style.css:1278]: Delete rather than comment out?

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by AleziaKurdis** [comment]
Added:
A group of 5 new **selection operations** related to **Hierarchy (Parent/Children)**

image

**Select Parent**: Select the immediate parent of the currently selected entities (discarding the initial selection)
**Select Top Parent**: Climb in the hierarchy over of the currently selected entities and Select the parent found at the top of the hierarchy (discarding the initial selection)
**Add Children To Selection**: Find the children recursively from the selected entities and add them to the current selection. (Initial selection stays included)
**Select Family**: Do  a  "Select Parent" followed by  a "Add Children To Selection". So from any entity, it selects its immediate parent and select  all the children under it.
**Select Top Family**: Do  a  "Select Top Parent" followed by  a "Add Children To Selection". So from any entity, it selects its Top parent and select all the children under it.

**"Inverse Selection"** is now working with **Ctrl-I**

Additional issues are planned to be addressed in the next days. Maybe wait before merging this PR.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5343935_15924f95.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5343935.tar_15924f95.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5343935.tar_15924f95.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5343935_15924f95.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5343935_15924f95.dmg

File: scripts/system/create/entityList/html/js/entityList.js
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/html/js/entityList.js:1568]: Do you need to test that only the control key is pressed and not Alt and/or Shift?

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js
Comment (ctrlaltdavid) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:715]:
```suggestion
        if (properties.parentID === Uuid.NULL) {
```

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js
Comment (ctrlaltdavid) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:726]:
```suggestion
                var childrenIDs = Entities.getChildrenIDs(that.selections[i]);
```
`childrenIDs` would be a nicer name, I think.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: scripts/system/create/entityList/html/js/entityList.js
Comment (AleziaKurdis) [reply scripts/system/create/entityList/html/js/entityList.js:1568]: Hmm. I'm not sure what you mean.
This want to respond to Ctrl-i for "Inverse Selection"
As what was already done for "Ctrl-a" (for "Select All") at line 1545
Please give me additional detail about what you would expect to see adjusted.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-3d3c98a.tar_bcebf362.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-3d3c98a_bcebf362.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-3d3c98a.tar_bcebf362.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-3d3c98a_bcebf362.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-3d3c98a_bcebf362.exe

File: scripts/system/create/entityList/html/js/entityList.js
Comment (ctrlaltdavid) [reply scripts/system/create/entityList/html/js/entityList.js:1568]: For example, test "if control key pressed but shift and alt keys aren't pressed".

Comment (AleziaKurdis) [reply scripts/system/create/entityList/html/js/entityList.js:1568]: Got it... current Ctrl-A but also Ctrl-Shift-A, Ctrl-Alt-A and Ctrl-Shift-Alt-A were also doing the Select All

I will change the 2 conditions like this:
if (controlKey && !shiftKey && !altKey && keyCodeString === "A") {
if (controlKey && !shiftKey && !altKey && keyCodeString === "I") {
so only CTRL-A and CTRL-I will do it.

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by AleziaKurdis** [comment]
Added 2 new optional columns to the Entity List (to address issue #673)
- **Creation Date**
- **Modification Date**
image

This was the last addition to this PR.
It can be merged once reviewed, tested and challenged if required.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5aa35e7.tar_8cc9c19b.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5aa35e7_8cc9c19b.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-5aa35e7.tar_8cc9c19b.gz

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-a099214.tar_a264fc35.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-a099214.tar_a264fc35.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-a099214_a264fc35.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-a099214_a264fc35.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-a099214_a264fc35.exe

File: scripts/system/create/entityList/entityList.js
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/entityList.js:239]:
```suggestion
                            lastEdited: formatToStringDateTime(properties.lastEdited)
```

File: scripts/system/create/entityList/entityList.js
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/entityList.js:262]:
```suggestion
        dateTime = dateTime + " " + zeroPad(d.getUTCHours(), 2) + ":" + zeroPad(d.getUTCMinutes(), 2) + ":" + zeroPad(d.getUTCSeconds(), 2); 
```

File: scripts/system/create/entityList/entityList.js
Comment (ctrlaltdavid) [inline scripts/system/create/entityList/entityList.js:269]:
```suggestion
        while (num.length < size) {
            num = "0" + num;
        }
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-d869b11_1e51df08.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-d869b11.tar_1e51df08.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-d869b11.tar_1e51df08.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-d869b11_1e51df08.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-d869b11_1e51df08.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-2a44068_1e51df08.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-2a44068.tar_1e51df08.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-2a44068.tar_1e51df08.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-2a44068_1e51df08.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR833/Vircadia-Alpha-PR833-2a44068_1e51df08.exe

**Comment by Aitolda** [comment]
Seems to be working as intended.  I do find myself missing the red delete button while in VR though.  Not a deal breaker, but an extra step.

**Comment by SilverfishVR** [comment]
Overall, I see no issues that prevents this from being merged, these are great additions so I say ship it 😄 

I do have comments though, these are related to VR and desktop equally.

* “Teleport to selected” should maybe be in actions instead of selection?

* If you leave either of the menus (actions, selection) open, ie clicking the button but not selecting an operation, it stays open even after disabling/enabling create mode, not a big problem but unintuitive.

* Sometimes the undo works weirdly after using other actions:
Make a cube, duplicate and move it.  Select the first cube and move your avatar away a bit. Use the “move selected entities to avatar” and then undo it. The moving of the duplicate cube will be undone instead of the “move selected entities to avatar” 

* I feel like it would be helpful if there was some visual indication of which entity is the active one (last selected) like you have in blender where selection order is also important, maybe a different selection color in the entity list?

* This may just be me, but from the names “Select family” and “select top family” it was not immediately clear what they do, I know from reading the PR but hmm.

* The optional entity list columns I have enabled do not persist between sessions, this was always the case but with these new columns it would be useful, e.g. I would want the Hierarchy to be enabled all the time because I really like that one.

**Comment by AleziaKurdis** [comment]
**Delete button:**
I do miss the presence of the Delete button either. I guess it's normal since we are used to it for years.
But the real question is:
Is the "Delete" is something we are doing more frequently than other actions?
Enough to justify having it exceptionally as a button?
If you think it is, then let me know. 
But if you can say the same thing for other actions, maybe it's why we can't continue to use a toolbar at the top.

**"Teleport to selection":**
It's kinda in a 3rd category. It's not a selection operation. it's not really an action (it does not alter the entities). It's more a "navigation". But currently, it's alone in this kind, so I wasn't considering to open a Navigation menu just for that.
So it can be put in any of these 2 menus. Or have a button in the toolbar (a button that won't have any tooltip to explain what it does.)
I put this as a question on discord. Let see what people will vote.

**Menu that stays open:**
This is not a classic dropdown. 
To be usable in HDM, it's more comfortable when the menu doesn't rely on any hover event to close.
Normally, you have no other choice than click outside to do anything else, which cancels the menu.
What is not a big deal, is that it doesn't close if you close the create app. 
This is a bug. I added it to my roadmap, I will address it with my the next loop of development. 
(I don't think it's critical enough to be addressed with this PR.)

**“Move Selected Entities to Avatar” is not recorded in history.**
Clearly a miss. (it's a brand new action, I didn't even think about this)
I added this to my roadmap, I will address it with my the next loop of development.
(I don't think this is a show stopper for this PR)

**Highlighting the last selected entity of a multiple selection.**
Nice idea. I added this to my roadmap. 
For the LIST, I will address it with my next loop of development.
For the in-world Selection, I will evaluate this, it's a bit more tricky. Possibly I would add a second box for the last selected item.

**"Select Family" and "Select Top Family"**
I agree, it might need other names.
I will open a question on Discord to get suggestions.

**Hierarchy column:** 
I would have set "Hierarchy" as active by default. But I didn't dare to impose it.
But I can do it if everyone agrees.

**Persisting column selection**
Something to consider. I will add this as an enhancement on my roadmap.

**Comment by kasenvr** [comment]
So, for now this seems to be good to merge, then follow-up with further rounds of PRs to address thoughts? :)

**Comment by AleziaKurdis** [comment]
That's the plan.

**Pull request merged**
